pkg/discovery/dtofactory/container_spec_entity_dto_builder.go

```diff
@@ -62,37 +61,42 @@ func (builder *containerSpecDTOBuilder) BuildDTOs() ([]*proto.EntityDTO, error)
-func (builder *containerSpecDTOBuilder) getCommoditiesSold(containerSpec *repository.ContainerSpec) ([]*proto.CommodityDTO, error) {
+func (builder *containerSpecDTOBuilder) getCommoditiesSold(containerSpecMetrics *repository.ContainerSpecMetrics) ([]*proto.CommodityDTO, error) {
 	var commoditiesSold []*proto.CommodityDTO
-	for _, commodityType := range ContainerSpecCommoditiesSold {
-		// commodities is a list of commodity DTOs of commodityType sold by container replicas represented by this
-		// ContainerSpec entity
-		commodities, exists := containerSpec.ContainerCommodities[commodityType]
+	for _, resourceType := range ContainerSpecResourceTypes {
+		commodityType, exist := rTypeMapping[resourceType]
+		if !exist {
+			glog.Errorf("Unsupported resource type %s when building commoditiesSold for ContainerSpec %s",
+				resourceType, containerSpecMetrics.ContainerSpecId)
+			continue
+		}
+		resourceMetrics, exists := containerSpecMetrics.ContainerMetrics[resourceType]
 		if !exists {
-			glog.V(4).Infof("ContainerSpec %s has no %s commodity from the collected ContainerSpec",
-				containerSpec.ContainerSpecId, commodityType)
+			glog.V(4).Infof("ContainerMetrics collected from ContainerSpec %s has no %s resource type",
+				containerSpecMetrics.ContainerSpecId, resourceType)
 			continue
 		}
 		commSoldBuilder := sdkbuilder.NewCommodityDTOBuilder(commodityType)
 
-		// Aggregate container replicas utilization data
-		utilizationDataPoints, err := builder.containerUtilizationDataAggregator.Aggregate(commodities)
+		// Aggregate container replicas utilization data.
+		// Note that the returned dataInterval is not the real sampling interval because there could be multiple data
+		// points from container replicas discovered at the same time in one set of data samples. This is just a calculated
+		// equivalent interval to be fed into percentile based algorithm in Turbo server side.
+		utilizationDataPoints, lastPointTimestamp, dataInterval, err := builder.containerUtilizationDataAggregator.Aggregate(resourceMetrics)
 		if err != nil {
-			glog.Errorf("Error aggregating commodity utilization data for ContainerSpec %s, %v",
-				containerSpec.ContainerSpecId, err)
+			glog.Errorf("Error aggregating %s utilization data for ContainerSpec %s, %v",
+				resourceType, containerSpecMetrics.ContainerSpecId, err)
 			continue
 		}
-		currentMs := time.Now().UnixNano() / int64(time.Millisecond)
-		// Currently we only collect one set of metrics data points from Kubernetes within a discovery cycle so the
-		// lastPointTimestampMs of utilizationData is current timestamp in milliseconds and intervalMs is 0.
-		commSoldBuilder.UtilizationData(utilizationDataPoints, currentMs, 0)
+		// Construct UtilizationData with multiple data points, last point timestamp in milliseconds and interval in milliseconds
+		commSoldBuilder.UtilizationData(utilizationDataPoints, lastPointTimestamp, dataInterval)
```

Comment:
Maybe we could talk about this, but if I'm reading the code correctly I don't think we are providing a good dataInterval. As I understand the platform's handling of the data interval, this value should be the time interval between individual samples in the utilization data array that we return in the discovery results. This means, for example, if we have 100 samples, a data interval of 10 minutes, and a last timestamp of 12 PM, the platform will interpret that to mean the last sample in the array is at 12 PM, the second to last sample is at 12 PM - 10 minutes, the third to last sample is at 12 PM - 20 minutes, etc. and the first sample is at 12 PM - 99 * 10 minutes. This means that the first data point will be over half a day before the first one, and I don't think that's what we want.

I think the data interval we actually want to set here is `dataIntervalMs / len(utilizationDataPoints)` so that, in the example above, the last data point will be interpreted as being at 12 PM and the first data point will be at 12 PM - 99 * (10 minutes / 100), or roughly 10 minutes before the last data point.

Reply:
Yue and I discussed. The data interval is actually correct. My understanding of it was based on an out-of-date comment in the code but the implementation is actually correct.